core/src/main/java/org/bitcoinj/core/TransactionInput.java

```diff
@@ -359,15 +359,29 @@ public void connect(TransactionOutput out) {
      * @return true if the disconnection took place, false if it was not connected.
      */
     public boolean disconnect() {
-        if (outpoint.fromTx == null) return false;
-        TransactionOutput output = outpoint.fromTx.getOutput((int) outpoint.getIndex());
-        if (output.getSpentBy() == this) {
-            output.markAsUnspent();
+        if (outpoint.fromTx != null) {
+            TransactionOutput output = outpoint.fromTx.getOutput((int) outpoint.getIndex());
             outpoint.fromTx = null;
-            return true;
+            if (output.getSpentBy() == this) {
```

Comment:
This part of code (up to line 370) could be extracted to the end of the method. Currently it's duplicated.